samples/matter/common/src/diagnostic/diagnostic_logs_provider.cpp

```diff
@@ -0,0 +1,164 @@
+	VerifyOrReturnError(freeSlot < kMaxLogSessionHandle, CHIP_ERROR_NO_MEMORY);
+
+	mIntentMap.Insert(freeSlot, std::move(intent));
+	outHandle = freeSlot;
+
+	ChipLogProgress(Zcl, "Starting Log collection for %s with session handle %d", GetIntentStr(intent), outHandle);
+
+	return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR DiagnosticLogProvider::EndLogCollection(LogSessionHandle sessionHandle)
+{
+	VerifyOrReturnError(sessionHandle < kMaxLogSessionHandle, CHIP_ERROR_NO_MEMORY);
+	VerifyOrReturnError(mIntentMap.Contains(sessionHandle), CHIP_ERROR_INTERNAL);
+
+	CHIP_ERROR err = CHIP_NO_ERROR;
+
+	IntentEnum intent = mIntentMap[sessionHandle];
+
+	/* Old intent is no needed anymore*/
+	mIntentMap.Erase(sessionHandle);
+
+	/* Do the specific action for the removing intent */
+	switch (intent) {
+	case IntentEnum::kEndUserSupport:
+		/* TODO: add support for End User Intent */
+		return CHIP_ERROR_NOT_IMPLEMENTED;
+	case IntentEnum::kNetworkDiag:
+		/* TODO: add support for End User Intent */
+		return CHIP_ERROR_NOT_IMPLEMENTED;
+	case IntentEnum::kCrashLogs:
+		err = ClearCrashLogs();
+		break;
+	default:
+		return CHIP_ERROR_INVALID_ARGUMENT;
+	}
+
+	ChipLogProgress(Zcl, "Ending Log collection for %s with session handle %d",
+			GetIntentStr(mIntentMap[sessionHandle]), sessionHandle);
+
+	return err;
+}
+
+CHIP_ERROR DiagnosticLogProvider::CollectLog(LogSessionHandle sessionHandle, MutableByteSpan &outBuffer,
+					     bool &outIsEndOfLog)
+{
+	VerifyOrReturnError(sessionHandle < kMaxLogSessionHandle, CHIP_ERROR_NO_MEMORY);
+	VerifyOrReturnError(mIntentMap.Contains(sessionHandle), CHIP_ERROR_INTERNAL);
+
+	CHIP_ERROR err = CHIP_NO_ERROR;
+
+	switch (mIntentMap[sessionHandle]) {
+	case IntentEnum::kEndUserSupport:
+		/* TODO: add support for End User Intent */
+		return CHIP_ERROR_NOT_IMPLEMENTED;
+	case IntentEnum::kNetworkDiag:
+		/* TODO: add support for Network Diag Intent */
+		return CHIP_ERROR_NOT_IMPLEMENTED;
+	case IntentEnum::kCrashLogs:
+		err = GetCrashLogs(outBuffer, outIsEndOfLog);
+		VerifyOrReturnError(CHIP_NO_ERROR == err, err, outBuffer.reduce_size(0));
+		break;
+	default:
+		return CHIP_ERROR_INVALID_ARGUMENT;
+	}
+
+	return CHIP_NO_ERROR;
+}
+
+size_t DiagnosticLogProvider::GetSizeForIntent(IntentEnum intent)
+{
+	/* TODO: Add implementation of this method */
+	return 0;
+}
+
+CHIP_ERROR DiagnosticLogProvider::GetLogForIntent(IntentEnum intent, MutableByteSpan &outBuffer,
+						  Optional<uint64_t> &outTimeStamp,
+						  Optional<uint64_t> &outTimeSinceBoot)
+{
+	LogSessionHandle sessionHandle = kInvalidLogSessionHandle;
+
+	CHIP_ERROR err = StartLogCollection(intent, sessionHandle, outTimeStamp, outTimeSinceBoot);
+	VerifyOrExit(CHIP_NO_ERROR == err, );
+
+	bool unusedOutIsEndOfLog;
+	err = CollectLog(sessionHandle, outBuffer, unusedOutIsEndOfLog);
+	VerifyOrExit(CHIP_NO_ERROR == err, );
+
+	err = EndLogCollection(sessionHandle);
+	VerifyOrExit(CHIP_NO_ERROR == err, );
+
+exit:
+	outBuffer.reduce_size(0);
```

Comment:
Should we do that in case of success? I think we don't want to set size to 0 in other case than `err != CHIP_NO_ERROR`.

Reply:
Oh, I forgot to push the recent changes